rbf-chunker/src/main.rs

```diff
@@ -0,0 +1,211 @@
+                    // Last packet was actually the footer.
+                    // Discard it.
+                    //
+                    // Wtf???
+                    eprintln!("{:?}: Discarding last packet as was footer.", filename);
+
+                    assert!(packet_number > 0);
+                    let filename = format!("packet{:05}", packet_number - 1);
+                    std::fs::remove_file(target_dir.join(filename))?;
+                    packet_infos.pop().unwrap();
+                }
+                File::create(target_dir.join("footer"))?
+                    .write_all(&working_buffer[..discard_so_far + read_so_far])?;
+                dump_packet_infos(&target_dir, &packet_infos)?;
+
+                return Ok(());
+            }
+            Ok(n) => {
+                let mut old_read_so_far = read_so_far;
+                read_so_far = read_so_far + n;
+                assert!(read_so_far <= PACKET_MAX_SIZE);
+                assert!(discard_so_far <= DISCARD_MAX_SIZE);
+
+                'try_again: loop {
+                    for end_read_so_far in old_read_so_far..read_so_far {
+                        if end_read_so_far < 3 {
+                            continue;
+                        }
+                        running_crc.update(&[working_buffer[discard_so_far + end_read_so_far - 3]]);
+                        let packet_info = PacketInfo::new(
+                            packet_number,
+                            packet_offset,
+                            discard_so_far,
+                            &working_buffer[discard_so_far..discard_so_far + end_read_so_far],
+                            running_crc.get(),
+                        );
+                        if end_read_so_far < PACKET_MIN_SIZE {
+                            continue;
+                        }
+                        if packet_info.crc == packet_info.target_crc && packet_info.crc != [0, 0] {
+                            let filename = format!("packet{:05}", packet_number);
+                            File::create(target_dir.join(filename))?.write_all(
+                                &working_buffer[discard_so_far..discard_so_far + end_read_so_far],
+                            )?;
+
+                            if discard_so_far != 0 {
+                                let filename = format!("packet{:05}-discard", packet_number);
+                                File::create(target_dir.join(filename))?
+                                    .write_all(&working_buffer[0..discard_so_far])?;
+                            }
+
+                            packet_infos.push(packet_info);
+                            if read_so_far != end_read_so_far {
+                                assert!(read_so_far > end_read_so_far);
+                                assert!(read_so_far <= PACKET_MAX_SIZE);
+                                assert!(discard_so_far <= DISCARD_MAX_SIZE);
+                                unsafe {
```

Comment:
I'm not `unsafe`-averse, but would prefer a `for` loop copy here. LLVM can turn it into a `memcpy` anyway.

Reply:
I'd argue that the for loop copy would make what the code is trying to do less obvious. The assertions above should insure this code never triggers UB.

Reply:
New version doesn't memcpy! whooo.